webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/write/PolicyWriterService.scala

```diff
@@ -305,168 +335,293 @@ class PolicyWriterServiceImpl(
+    /*
+     * Idea: sort template by content (ie what is cached). Put each unique cache in a seq with the corresponding list of
+     * template to fill.
+     * Get the the corresponding list of filled template.
+     * Write it.
+     * No semaphore, no synchronisation.
+     */
+
+    // template are created based on "content"
+    val byTemplate = (for {
+      p  <- preparedTemplates
+      pt <- p.preparedTechniques
+      t  <- pt.templatesToProcess
+    } yield {
+      (t.content, TemplateFillInfo(t.id, t.destination, p.paths.newFolder, pt.environmentVariables, pt.reportIdToReplace))
+    }).groupMap(_._1)(_._2)
+
+
+    import org.antlr.stringtemplate.StringTemplate
+    import com.normation.stringtemplate.language.NormationAmpersandTemplateLexer
+
+    // now process by template, which are not thread safe but here, accessed sequentially
+    parrallelSequence(byTemplate.toSeq) { case (content, seqInfos) =>
+      for {
+        t0      <- currentTimeNanos
+        parsed  <- IOResult.effect(s"Error when trying to parse template '${seqInfos.head.destination}'") { // head ok because of groupBy
+                     new StringTemplate(content, classOf[NormationAmpersandTemplateLexer])
+                   }
+        t1      <- currentTimeNanos
+        _       <- fillTimer.get.update(_ + t1 - t0)
+        _       <- ZIO.traverse_(seqInfos) { info =>
+                     for {
+                       // we need to for {} yield {} to free resources
+                       replacedDest   <- for {
+                                           _      <- PolicyGenerationLoggerPure.trace(s"Create policies file ${info.newFolder} ${info.destination}")
+                                           t0     <- currentTimeNanos
+                                           filled <- FillTemplateThreadUnsafe.fill(info.destination, parsed, info.envVars, fillTimer, info.reportIdToReplace.map(id => (Policy.TAG_OF_RUDDER_MULTI_POLICY, id.getRudderUniqueId)))
+                                           t1     <- currentTimeNanos
+                                           _      <- writeTimer.fillTemplate.update(_ + t1 - t0)
+                                         } yield {
+                                           filled
+                                         }
+                       (replaced, dest) = replacedDest
+                       t1               <- currentTimeNanos
+                       _                <- File(info.newFolder, dest).createParentsAndWrite(replaced).chainError(
+                                               s"Bad format in Technique ${info.id.toString} (file: ${info.destination})"
+                                           )
+                       t2               <- currentTimeNanos
+                       _                <- writeTimer.writeTemplate.update(_ + t2 - t1)
+                     } yield ()
```

Comment:
I am a bit scared by the potential memory usage here, as we keep a lot of data in memory
I'm going to accept it, because it seems ok on our test platform, but let's keep an eye on this